server/src/main/java/com/thoughtworks/go/server/service/AgentService.java

```diff
@@ -126,143 +139,412 @@ public AgentsViewModel registeredAgents() {
-        }
-        return agents;
+    private AgentsViewModel toAgentViewModels(AgentInstances agentInstances) {
+        return stream(agentInstances.spliterator(), false)
+                    .map(this::toAgentViewModel)
+                    .collect(Collectors.toCollection(AgentsViewModel::new));
     }
 
     private AgentViewModel toAgentViewModel(AgentInstance instance) {
-        return new AgentViewModel(instance, environmentConfigService.environmentsFor(instance.getUuid()));
+        return new AgentViewModel(instance, instance.getAgent().getEnvironmentsAsList());
     }
 
     public AgentInstances findRegisteredAgents() {
         return agentInstances.findRegisteredAgents();
     }
 
-    private boolean isUnknownAgent(AgentInstance agentInstance, OperationResult operationResult) {
-        if (agentInstance.isNullAgent()) {
-            String agentNotFoundMessage = String.format("Agent '%s' not found", agentInstance.getUuid());
-            operationResult.notFound("Agent not found.", agentNotFoundMessage, HealthStateType.general(HealthStateScope.GLOBAL));
+    private boolean agentNotFound(Agent agent, OperationResult result) {
+        if (agent.isNull()) {
+            String agentNotFoundMessage = format("Agent '%s' not found", agent.getUuid());
+            result.notFound("Agent not found.", agentNotFoundMessage, general(GLOBAL));
             return true;
         }
+
         return false;
     }
 
-    private boolean hasOperatePermission(Username username, OperationResult operationResult) {
+    private boolean doesNotHaveOperatePermission(Username username, OperationResult result) {
         if (!securityService.hasOperatePermissionForAgents(username)) {
             String message = "Unauthorized to operate on agent";
-            operationResult.forbidden(message, message, HealthStateType.general(HealthStateScope.GLOBAL));
-            return false;
+            result.forbidden(message, message, general(GLOBAL));
+            return true;
         }
-        return true;
+        return false;
     }
 
-    public AgentInstance updateAgentAttributes(Username username, HttpOperationResult result, String uuid, String newHostname, String resources, String environments, TriState enable) {
-        if (!hasOperatePermission(username, result)) {
+    public AgentInstance updateAgentAttributes(Username username, HttpOperationResult result, String uuid,
+                                               String hostname, String resources, String environments,
+                                               TriState state) {
+        AgentUpdateValidator validator = new AgentUpdateValidator(username, agentInstances.findAgent(uuid), hostname, environments,
+                                                                  resources, state, result, goConfigService);
+        Agent agent = agentDao.agentByUuid(uuid);
+        try {
+            if(validator.canContinue()) {
+                validator.validate();
```

Comment:
`validator.validate()` will throw a `RecordNotFoundException` if an agent does not exist. This will be handled by the catch block for all exceptions and would return a `500`. Guess a test for these scenarios will help.

Reply:
Added a specific catch for `RecordNotFoundException` which will not update the `HttpOperationResult` object. Tests have been added.